django_log_formatter_asim/events.py

```diff
@@ -0,0 +1,212 @@
+    :param severity: Optional severity of the event, defaults to "Informational", otherwise one of:
+                        - "Informational"
+                        - "Low"
+                        - "Medium"
+                        - "High"
+    :param time_generated: Optional datetime for when the event happened, otherwise datetime.now
+    :param result_details: Optional string describing any details associated with the events outcome.
+                           This field is typically populated when the result is a failure.
+    :param message: Optional string describing the reason why the log was generated.
+
+    See also: https://learn.microsoft.com/en-us/azure/sentinel/normalization-schema-authentication
+    """
+    if user == None:
+        user = {}
+    if server == None:
+        server = {}
+    if client == None:
+        client = {}
+
+    event_created = time_generated or datetime.datetime.now(tz=datetime.timezone.utc)
+
+    event = {
+        "EventCreated": event_created.isoformat(),  # TODO: Should this really be EventCreated, or TimeGenerated
+        "EventSeverity": severity or _default_severity(result),
+        "EventOriginalType": _event_code(type, result),
+        "EventType": type,
+        "EventResult": result,
+        "LogonMethod": login_method,
+        "EventSchema": "Authentication",
+        "EventSchemaVersion": "0.1.4",
+    }
+
+    if "hostname" in server:
+        event["DvcHostname"] = server["hostname"]
+    elif hasattr(request, "environ") and "SERVER_NAME" in request.environ:
+        event["DvcHostname"] = request.environ["SERVER_NAME"]
+
+    if "ip_address" in client:
+        event["SrcIpAddr"] = client["ip_address"]
+    elif hasattr(request, "environ") and "REMOTE_ADDR" in request.environ:
+        event["SrcIpAddr"] = request.environ.get("REMOTE_ADDR")
+
+    if "role" in user:
+        event["ActorUserType"] = user["role"]
+
+    if "sessionId" in user:
+        event["ActorSessionId"] = user["sessionId"]
+    elif request.session.session_key:
+        event["ActorSessionId"] = request.session.session_key
+
+    if "username" in user:
+        event["ActorUserName"] = user["username"]
+    elif request.user.username:
+        event["ActorUserName"] = request.user.username
+
+    if result_details:
+        event["EventResultDetails"] = result_details
+
+    if message:
+        event["EventMessage"] = message
+
+    if "ip_address" in server:
+        event["DvcIpAddr"] = server["ip_address"]
+
+    sys.stdout.write(json.dumps(event) + "\n")
+    sys.stdout.flush()
+
+
+log_authentication.Type = AuthenticationType
+log_authentication.Result = AuthenticationResult
+log_authentication.LoginMethod = AuthenticationLoginMethod
+log_authentication.Severity = Severity
+
+
+def _default_severity(result):
+    return Severity.Informational if result == AuthenticationResult.Success else Severity.Medium
+
+
+def _event_code(type, result):
```

Comment:
Type hints on these?
```suggestion
def _event_code(type: AuthenticationType, result: AuthenticationResult) -> str:
```

Reply:
Not a biggie, but `type` clashes with a builtin. Won't cause any issues, but may upset IDEs. Could be auth_type or something.

Reply:
Great spot! Had a good old think about this, and thought that the name `event` was a better term for the variable containing `Login` or `Logoff`. Has resulted in a bunch of renames everywhere, but I think it's a clearer term. Especially as this name forms part of the external API that is being called.


```python
log_authentication(
    wsgi_request,
    event=log_authentication.Event.Logoff,
    result=log_authentication.Result.Success,
    login_method=log_authentication.LoginMethod.UsernamePassword,
)
```